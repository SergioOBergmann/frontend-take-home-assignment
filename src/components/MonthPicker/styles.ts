import styled from 'styled-components';

const ArrowButton = styled.button`
  width: 48px;
  height: 56px;

  background: #e1e8ed;
  border: 1px solid transparent;
`;

const ArrowButtonSpan = styled.span`
  width: 11.313708px;
  height: 11.313708px;

  border: solid #657786;
  display: inline-block;
`;

export const PreviousButton = styled(ArrowButton as any)`
  border-radius: 4px 0px 0px 4px;
`;

export const PreviousButtonSpan = styled(ArrowButtonSpan as any)`
  margin-left: 8.8px;

  border-width: 0 0 2px 2px;
  transform: rotate(45deg);
`;

export const Content = styled.div`
  flex-direction: column;
  flex: 1;
  width: 100%;

  background: #ffffff;
  border: 1px solid #e1e8ed;
  box-sizing: border-box;
`;

export const MonthSpan = styled.span`
  margin-top: 6px;

  font-weight: 600;
  font-size: 20px;

  text-align: center;
  letter-spacing: -0.166667px;
`;

export const YearSpan = styled.span`
  font-size: 16px;
  line-height: 20px;

  text-align: center;
`;

export const NextButton = styled(ArrowButton as any)`
  border-radius: 0px 4px 4px 0px;
`;

export const NextButtonSpan = styled(ArrowButtonSpan as any)`
  margin-right: 4.8px;

  border-width: 2px 0 0 2px;
  transform: rotate(135deg);
`;
