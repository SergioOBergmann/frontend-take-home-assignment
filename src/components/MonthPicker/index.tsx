import * as React from 'react';

import { InputDiv, InputLabel, LOCALE } from '../../styles/global';

import {
  PreviousButton,
  PreviousButtonSpan,
  Content,
  MonthSpan,
  YearSpan,
  NextButton,
  NextButtonSpan
} from './styles';

export interface MonthDate {
  month: number;
  year: number;
}

type MonthPickerProps = {
  label: string;
  minDate: MonthDate;
  step?: number;
  defaultDate?: MonthDate;
  onChange: (date: MonthDate) => void;
};

const MonthPicker = ({
  label,
  minDate,
  step = 1,
  defaultDate = minDate,
  onChange
}: MonthPickerProps): JSX.Element => {
  const [date, setDate] = React.useState(defaultDate);

  const getMonthFormattedValue = () => {
    const dateFull = new Date(date.year, date.month - 1, 1);
    return dateFull.toLocaleString(LOCALE, {
      month: 'long'
    });
  };

  const handleChange = (value: number) => {
    const dateFull = new Date(date.year, date.month - 1 + value * step, 1);

    const minDateFull = new Date(minDate.year, minDate.month - 1, 1);

    if (value >= 0 || dateFull >= minDateFull) {
      const newDate = {
        month: dateFull.getMonth() + 1,
        year: dateFull.getFullYear()
      };
      setDate(newDate);
      onChange(newDate);
    }
  };

  const handleKeyDown = (event: KeyboardEvent) => {
    if ((event.currentTarget as Document).activeElement.localName !== 'input') {
      if (event.key === 'ArrowRight') {
        handleChange(1);
      } else if (event.key === 'ArrowLeft') {
        handleChange(-1);
      }
    }
    document.removeEventListener('keydown', handleKeyDown);
  };

  const handleKeyUp = (event: KeyboardEvent) => {
    document.addEventListener('keydown', handleKeyDown, { once: true });
  };

  React.useEffect(() => {
    document.addEventListener('keydown', handleKeyDown);
    document.addEventListener('keyup', handleKeyUp);
    return function cleanup() {
      document.removeEventListener('keydown', handleKeyDown);
      document.removeEventListener('keyup', handleKeyUp);
    };
  });

  return (
    <>
      <InputDiv>
        <InputLabel>{label}</InputLabel>
        <PreviousButton
          aria-label="Previous month"
          onClick={() => handleChange(-1)}
        >
          <PreviousButtonSpan />
        </PreviousButton>
        <Content>
          <MonthSpan>{getMonthFormattedValue()}</MonthSpan>
          <YearSpan>{date.year}</YearSpan>
        </Content>
        <NextButton aria-label="Next month" onClick={() => handleChange(1)}>
          <NextButtonSpan />
        </NextButton>
      </InputDiv>
    </>
  );
};

export default MonthPicker;
