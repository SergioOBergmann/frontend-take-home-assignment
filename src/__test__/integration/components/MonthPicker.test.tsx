import * as React from 'react';

import {
  render,
  RenderResult,
  fireEvent,
  screen
} from '@testing-library/react';

import MonthPicker from '../../../components/MonthPicker';

let documentBody: RenderResult;

describe('<MonthPicker />', () => {
  let onChange = jest.fn();
  beforeEach(() => {
    onChange = jest.fn();
    documentBody = render(
      <MonthPicker
        label="Reach goal by"
        defaultDate={{
          year: new Date().getFullYear(),
          month: new Date().getMonth() + 3
        }}
        minDate={{
          year: new Date().getFullYear(),
          month: new Date().getMonth() + 2
        }}
        onChange={onChange}
      />
    );
  });

  it('matches snapshot', () => {
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  describe('calls onPreviousButtonClicked prop', () => {
    it('when ArrowLeft key down', () => {
      fireEvent.keyDown(document, { key: 'ArrowLeft', code: 'ArrowLeft' });
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('when PreviousButton clicked', () => {
      fireEvent.click(screen.getByRole('button', { name: /Previous month/i }));
      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  describe('calls onNextButtonClicked prop', () => {
    it('when ArrowRight key down', () => {
      fireEvent.keyDown(document, { key: 'ArrowRight', code: 'ArrowRight' });
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('when NextButton clicked', () => {
      fireEvent.click(screen.getByRole('button', { name: /Next month/i }));
      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  it('no calls when any different key down and any key up', () => {
    fireEvent.keyDown(document, { key: 'ArrowUp', code: 'ArrowUp' });
    expect(onChange).not.toHaveBeenCalled();

    fireEvent.keyDown(document, { key: 'ArrowDown', code: 'ArrowDown' });
    expect(onChange).not.toHaveBeenCalled();

    fireEvent.keyUp(document, { key: 'ArrowRight', code: 'ArrowRight' });
    expect(onChange).not.toHaveBeenCalled();

    fireEvent.keyUp(document, { key: 'ArrowLeft', code: 'ArrowLeft' });
    expect(onChange).not.toHaveBeenCalled();
  });
});
