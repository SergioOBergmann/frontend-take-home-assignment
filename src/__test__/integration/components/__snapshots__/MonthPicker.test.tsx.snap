// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MonthPicker /> matches snapshot 1`] = `
<body>
  <div>
    <div
      class="sc-ifAKCX drfbAu"
    >
      <label
        class="sc-EHOje hUHzcE"
      >
        Reach goal by
      </label>
      <button
        aria-label="Previous month"
        class="sc-gzVnrw sc-dnqmqq ejpLKN"
      >
        <span
          class="sc-htoDjs sc-iwsKbI gEcacA"
        />
      </button>
      <div
        class="sc-gZMcBi jzBwSO"
      >
        <span
          class="sc-gqjmRU jrJxfM"
        >
          December
        </span>
        <span
          class="sc-VigVT cfNeXY"
        >
          2020
        </span>
      </div>
      <button
        aria-label="Next month"
        class="sc-gzVnrw sc-jTzLTM YEbUE"
      >
        <span
          class="sc-htoDjs sc-fjdhpX cpRatT"
        />
      </button>
    </div>
  </div>
</body>
`;
